tlatools/org.lamport.tlatools/src/tla2sany/semantic/Errors.java

```diff
@@ -71,58 +83,124 @@ public boolean equals(Object o) {
+  }
+
+  /**
+   * Append a message to the log. If location is null, the value
+   * {@link Location.nullLoc} is assigned. Idempotent; will not append the
+   * same message to the log multiple times. Returns {@link AbortException}
+   * which can optionally be thrown by the caller, if error is fatal.
+   *
+   * @param code The standardized error code associated with the message.
+   * @param loc A spec location associated with the message.
+   * @param str A human-readable message.
+   * @return An exception which can optionally be thrown by the caller.
+   */
+  public final AbortException addMessage(ErrorCode code, Location loc, String str) {
+    loc = null == loc ? Location.nullLoc : loc;
+    final ErrorDetails message = new ErrorDetails(code, loc, str);
+    if (!this.messages.contains(message)) {
+      this.messages.add(message);
     }
-    
-    return new AbortException(error, this);
+
+    return new AbortException(message, this);
   }
 
-  public final boolean isSuccess()             { return this.errors.isEmpty(); }
+  public final boolean isSuccess() {
+    return this.getMessagesOfAtLeastLevel(ErrorLevel.ERROR).isEmpty();
+  }
 
-  public final boolean isFailure()             { return !this.isSuccess(); }
+  public final boolean isFailure() {
+    return !this.isSuccess();
+  }
 
-  public final int     getNumErrors()          { return this.errors.size(); }
+  public final int getNumErrors() {
+    return this.getErrorDetails().size();
+  }
 
-  public final int     getNumMessages()        { return this.errors.size() + this.warnings.size(); }
+  public final int getNumMessages() {
+    return this.messages.size();
+  }
 
-  public final String  toString()  {
+  public final String toString()  {
     StringBuffer ret = new StringBuffer("");
 
-    ret.append((this.errors.size() > 0) ? "*** Errors: " + this.errors.size() + "\n\n" : "");
-    for (final ErrorDetails error : this.errors)   {
+    final List<ErrorDetails> errors = this.getErrorDetails();
+    ret.append((errors.size() > 0) ? "*** Errors: " + errors.size() + "\n\n" : "");
+    for (final ErrorDetails error : errors)   {
       ret.append(error.toString() + "\n\n\n");
     }
 
-    ret.append((this.warnings.size() > 0) ? "*** Warnings: " + this.warnings.size() + "\n\n" : "");
-    for (final ErrorDetails error : this.warnings) {
+    final List<ErrorDetails> warnings = this.getWarningDetails();
+    ret.append((warnings.size() > 0) ? "*** Warnings: " + warnings.size() + "\n\n" : "");
+    for (final ErrorDetails error : warnings) {
+      ret.append(error.toString() + "\n\n\n");
+    }
+
+    final List<ErrorDetails> lint = this.getMessagesOfLevel(ErrorLevel.LINT);
+    ret.append((lint.size() > 0) ? "*** Lint: " + lint.size() + "\n\n" : "");
+    for (final ErrorDetails error : lint) {
       ret.append(error.toString() + "\n\n\n");
     }
 
     return ret.toString();
   }
+
+  private static boolean atLeastLevel(ErrorCode code, ErrorLevel base) {
```

Comment:
`ErrorLevel` appears to assume a total order. Are SANY messages actually totally ordered, or is `ErrorLevel` functioning more like an `ErrorCategory`, where values represent categories with no inherent or only partial ordering? Additionally, there seems to be some conceptual overlap with `severity`.

Reply:
It is a total order. This will eventually be the basis for a --logLevel and --errorLevel CLI parameter to control output & what is interpreted as an error.

Reply:
Every logging system I know of manages to use a linear order of log levels. I don't see the purpose of categorizing messages in a non-linear hierarchy. What would be the use-case of that?

Reply:
Use-case: Heuristic-based vs deterministic-based warnings

Reply:
A use case is like "user will be able treat Lint, Warning, and Error level messages as errors (output to stderr and return nonzero exit code) using the command line parameter `-errorLevel lint`. What is a comparable use case for heuristic-based vs. deterministic-based warnings?

Reply:
It seems preferential to conform to common user experience conventions as much as possible. Software engineer expectation about tool behavior & parameters is a valuable repository of existing knowledge we can rely on instead of having to re-educate people about how our tool is special.